File: apps/backend/src/portfolio/PortfolioAsset.ts
import {
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { Portfolio } from './Portfolio';
import { AssetTradingRecord } from './AssetTradingRecord';
import { AssetDividendRecord } from './AssetDividendRecord';
import { UUID } from '@mid/shared';

@Entity()
export class PortfolioAsset {
  @PrimaryGeneratedColumn('uuid')
  id: UUID;

  @ManyToOne(() => Portfolio, (portfolio) => portfolio.portfolioAssets, {
    lazy: true,
    nullable: false,
  })
  @JoinColumn({ name: 'portfolio_id' })
  portfolio: Portfolio;

  @Column({
    type: 'varchar',
    nullable: false,
  })
  intendedAsset: string;

  @Column({
    type: 'numeric',
    nullable: false,
  })
  targetRatio: number;

  @OneToMany(
    () => AssetTradingRecord,
    (tradingRecord) => tradingRecord.portfolioAsset,
    {
      lazy: true,
      cascade: true,
      orphanedRowAction: 'delete',
    },
  )
  records: AssetTradingRecord[];

  @OneToMany(
    () => AssetDividendRecord,
    (dividendRecord) => dividendRecord.portfolioAsset,
    {
      lazy: true,
      cascade: true,
      orphanedRowAction: 'delete',
    },
  )
  dividends: AssetDividendRecord[];
}

File: backend/src/main/kotlin/com/kyoulho/mid/portfolio/PortfolioAsset.kt
package com.kyoulho.mid.portfolio

import jakarta.persistence.*

@Entity
data class PortfolioAsset(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    val id: String? = null,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "portfolio_id", nullable = false)
    val portfolio: Portfolio,

    @Column(nullable = false)
    val intendedAsset: String,

    @Column(nullable = false)
    val targetRatio: Double,

    @OneToMany(mappedBy = "portfolioAsset", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.LAZY)
    val records: List<AssetTradingRecord> = mutableListOf(),

    @OneToMany(mappedBy = "portfolioAsset", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.LAZY)
    val dividends: List<AssetDividendRecord> = mutableListOf()
)
